input/elasticapm/processor.go

```diff
@@ -363,7 +364,12 @@ func (p *Processor) handleStream(
 
 // processBatch processes the batch and returns it to the pool after it's been processed.
 func (p *Processor) processBatch(ctx context.Context, processor modelpb.BatchProcessor, batch *modelpb.Batch) error {
-	defer p.batchPool.Put(batch)
+	defer func() {
+		for i := range *batch {
+			(*batch)[i] = nil
+		}
+		batchPool.Put(batch)
+	}()
```

Comment:
Should we update apm-data to use vtproto's "pool" feature, and then use ReturnToVTPool/*ProtoFromVTPool functions? ReturnToVTPool will also take care of resetting.

Reply:
I just noticed today that this was missing from the current generation. It would be good to add it back.

Reply:
> Should we update apm-data to use vtproto's "pool" feature, and then use ReturnToVTPool/*ProtoFromVTPool functions? ReturnToVTPool will also take care of resetting.

Could you clarify ? AFAICS `modelpb.Batch` is not a protobuf message, just a type alias, so it can't use the protobuf generated methods.

> I just noticed today that this was missing from the current generation. It would be good to add it back.

That's strange, the pool feature is enabled even though not used. Thanks for spotting that!

Reply:
The structs need to be specified explicitly. I can add it back later if that's OK with you.

Reply:
We could potentially define a Batch ourselves:

```proto
message Batch {
  repeated APMEvent events = 1;
}
```

And let the pool plugin do its thing

Reply:
I can look into it, my comment was mostly due to surprising behaviour in the different usage of the flag.

> We could potentially define a Batch ourselves:
> 
> message Batch {
>   repeated APMEvent events = 1;
> }
> 
> And let the pool plugin do its thing

That would generate a struct but we're not really interested in reusing protobuf models (not in this case at least), we want to reuse the slice to avoid allocating a new one for each batch.

Reply:
If defining a Batch message type doesn't work out, then how about adding a method to Batch ourselves?

```go
// Reset resets each event in the batch, and truncates the slice to zero length.
func (b *Batch) Reset() {
    for _, event := range *b {
	event.ResetVT()
    }
    *b = (*b)[:0]
}
```

Then

```go
defer batchPool.Put(batch)
defer batch.Reset()
```

Reply:
I guess that doesn't help with returning the events to the pool though...

Reply:
> ```go
> // Reset resets each event in the batch, and truncates the slice to zero length.
> func (b *Batch) Reset() {
>     for _, event := range *b {
> 	event.ResetVT()
>     }
>     *b = (*b)[:0]
> }
> ```

I might be missing something, why do we want to use a pool of `APMEvent`s ? We're not reusing protobuf models and I don't think that's possible: we rely on the baseevent from APM server which is then cloned before being used by the decoder.

Reply:
APMEvents will be more expensive to allocate than the slice. Good point about the base events. I think we may need to move away from this approach to enable pooling. We could use a pool and `proto.Merge` the base event on top of an empty event, but reflection will kill us.

Anyway, seems like this is not straightforward at the moment. Let's come back to pooling the events later.

Reply:
I must admin I was under the impression that the continuously allocating the slice would be more expensive.

Agree with what you said, let's come back to this later :+1: